libp2p/switch.nim

```diff
@@ -211,12 +213,13 @@ proc upgradeOutgoing(s: Switch, conn: Connection): Future[Connection] {.async, g
     raise newException(UpgradeFailedError,
       "No peerInfo for connection, stopping upgrade")
 
+  s.connManager.updateConn(conn, sconn)
```

Comment:
isn't there a risk here that the un-upgraded connection gets used in a separate call to connmanager, due to async reordering?

Reply:
ie why are connections that are not upgraded yet held in the same connection manager structure? it seems that they should only be allowed in there _after_ they've gone through upgrade?

Reply:
> isn't there a risk here that the un-upgraded connection gets used in a separate call to connmanager, due to async reordering?

It shouldn't and if it does then that is a logical bug, not a concurrency one. If such a bug is introduced it can only be due to the update being called at the wrong place or for the wrong connection.

> ie why are connections that are not upgraded yet held in the same connection manager structure? it seems that they should only be allowed in there after they've gone through upgrade?

I've partially addressed this in https://github.com/status-im/nim-libp2p/pull/384#discussion_r498884362, but to reiterate, `ConnManager` is intended to be the singly source of truth in terms of managing connection related resource. By consolidating all connection tracking logic in one place, we're making it easier to maintain global counts, per peer counts and eventually, add heuristics as to prioritize connections, apply throttle, etc... 

To address the specific point of "unidentified" connections, if we don't track them in the connection manager and we still want to account for them (in-flight connection), that logic would end up being spread across different modules. Which is what I initially attempted to do, but quickly realized that I'm going to have to share the semaphore across those modules that will create or upgrade connections, such as the switch (for now and the transports themself in a further refactor).

Another point to keep in mind is that the upgrade logic is moving down to each individual transport and when that happens and transports like `websockets`, `webrtc` and in particular `quic` are introduced, they will all "produce" connections in different ways and at different stages, having a connection manager that is capable of working across all those different flows is very handy.

Reply:
What I don't understand is the following:

1. `conn` is added to connectionmanager by task A - for example through a dial where the peerid was known
2. `conn` starts going through upgrade process which involves several negotiation steps, so task A hands over execution to other tasks
3. task B wants a connection to the same peer, and finds  it in connmanager - notably, before upgrade has finished.

what prevents the above scenario?

Reply:
> What I don't understand is the following:
> 
> 1. `conn` is added to connectionmanager by task A - for example through a dial where the peerid was known
> 2. `conn` starts going through upgrade process which involves several negotiation steps, so task A hands over execution to other tasks
> 3. task B wants a connection to the same peer, and finds  it in connmanager - notably, before upgrade has finished.
> 
> what prevents `conn` from being given to task B the above scenario?

The important thing to keep in mind is that I haven't really changed the existing flow all that much and things still behave very similarily to what they did prior to this changes.

- For dialing (outgoing), we hold a lock in the dial itself that prevents other dials to happen concurrently until the upgrade has finished - https://github.com/status-im/nim-libp2p/blob/tcp-limits2/libp2p/switch.nim#L278-L280.
  - I've removed the usage of `selectConn ` and instead we call [`getStream(peerId)`](https://github.com/status-im/nim-libp2p/blob/tcp-limits2/libp2p/switch.nim#L283), which won't return a stream unless the connection has upgraded (in the case of TCP, up until it got a muxer attached). This prevents raise conditions related to working directly with the connection in the Switch and instead moves that flow into the `ConnManager`
  - The requirement to keep only one outgoing connection is still in place, as we only allow to dial when we know the `peerId` (this might change in the future, but this is how it works right now).

- For incoming connections, we allow several concurrent connections to exist at the same time. However, the same logic as with the dial flow applies, `getStream` won't return anything until a muxer has been attached, which will only happen in [`muxerHandler`](https://github.com/status-im/nim-libp2p/blob/tcp-limits2/libp2p/switch.nim#L503). If a dial is initiated, it will either get a stream from an existing incoming connection, or attempt to establish a new one according to the dial flow.

This changes try hard to not modify the existing flow, rather they build on top of what's there already. This is by design and it will evolve a bit more as I rework the upgrade flow further, but right now I want to scope things as much as possible.